### ISSUE TITLE
KCP upgrade fails and pod in CrashLoopBackOff due nil pointer error

### ISSUE THREAD

**Issue by Xenwar** [kubernetes-sigs/cluster-api#3363]
**What steps did you take and what happened:**
Tried to upgrade a KCP node

**What did you expect to happen:**
That upgrade succeeds. 

**Anything else you would like to add:**
The main focus of this issue is handling the **nil pointer error**

**Environment:**

- Cluster-api version: v0.3.0
- Minikube version: v1.12.0
- Kubernetes version: (use `kubectl version`): v1.18.3
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by Xenwar** [comment]
```
ubuntu@upgrade-env-anwar:~$ kubectl logs capi-kubeadm-control-plane-controller-manager-7c8ccbb5f-gvmt8 -n capi-kubeadm-control-plane-system manager
I0717 13:13:34.352148       1 listener.go:40] controller-runtime/metrics "msg"="metrics server is starting to listen"  "addr"="127.0.0.1:8080"
I0717 13:13:34.353529       1 main.go:119] setup "msg"="starting manager"  
I0717 13:13:34.355842       1 leaderelection.go:242] attempting to acquire leader lease  capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi...
I0717 13:13:34.358014       1 internal.go:356] controller-runtime/manager "msg"="starting metrics server"  "path"="/metrics"
I0717 13:13:51.807949       1 leaderelection.go:252] successfully acquired lease capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi
I0717 13:13:51.814285       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"version":"","infrastructureTemplate":{},"kubeadmConfigSpec":{}},"status":{"initialized":false,"ready":false}}}
I0717 13:13:51.921542       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"clusterName":"","bootstrap":{},"infrastructureRef":{}},"status":{"bootstrapReady":false,"infrastructureReady":false}}}
I0717 13:13:52.025995       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"controlPlaneEndpoint":{"host":"","port":0}},"status":{"infrastructureReady":false,"controlPlaneInitialized":false}}}
I0717 13:13:52.131970       1 controller.go:171] controller-runtime/controller "msg"="Starting Controller"  "controller"="kubeadmcontrolplane"
I0717 13:13:52.132297       1 controller.go:190] controller-runtime/controller "msg"="Starting workers"  "controller"="kubeadmcontrolplane" "worker count"=1
I0717 13:13:52.352864       1 kubeadm_control_plane_controller.go:252] controllers/KubeadmControlPlane "msg"="Upgrading Control Plane" "cluster"="test1" "kubeadmControlPlane"="test1" "namespace"="metal3" 
E0717 13:13:53.565171       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 253 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x16a8280, 0x26f9f60)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x16a8280, 0x26f9f60)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc000544300, 0x1b02b40, 0xc000058068, 0xc00010b980, 0xc000512f00, 0xc00039bf20, 0xc0005dc000, 0xc0000f5800, 0x1, 0x1, ...)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:376 +0x5ff
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc000544300, 0x1b02b40, 0xc000058068, 0xc00010b980, 0xc000512f00, 0x0, 0x0, 0x0, 0xc0004b1330)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:253 +0x1250
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc000544300, 0xc0004b1330, 0x6, 0xc0004b1320, 0x5, 0xc000611c00, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:190 +0x6a3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001dc480, 0x1709460, 0xc00046f220, 0x440500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001dc480, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001dc480)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000186ca0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000186ca0, 0x3b9aca00, 0x0, 0x1, 0xc00044be00)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000186ca0, 0x3b9aca00, 0xc00044be00)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x152ca2f]

goroutine 253 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:55 +0x105
panic(0x16a8280, 0x26f9f60)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc000544300, 0x1b02b40, 0xc000058068, 0xc00010b980, 0xc000512f00, 0xc00039bf20, 0xc0005dc000, 0xc0000f5800, 0x1, 0x1, ...)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:376 +0x5ff
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc000544300, 0x1b02b40, 0xc000058068, 0xc00010b980, 0xc000512f00, 0x0, 0x0, 0x0, 0xc0004b1330)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:253 +0x1250
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc000544300, 0xc0004b1330, 0x6, 0xc0004b1320, 0x5, 0xc000611c00, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:190 +0x6a3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001dc480, 0x1709460, 0xc00046f220, 0x440500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001dc480, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001dc480)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000186ca0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000186ca0, 0x3b9aca00, 0x0, 0x1, 0xc00044be00)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000186ca0, 0x3b9aca00, 0xc00044be00)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:193 +0x328
ubuntu@upgrade-env-anwar:~$ 

```

**Comment by Xenwar** [comment]
```
buntu@upgrade-env-anwar:~$ kubectl get pods -A
NAMESPACE                           NAME                                                            READY   STATUS             RESTARTS   AGE
capi-kubeadm-bootstrap-system       capi-kubeadm-bootstrap-controller-manager-c4786d5f-zln7g        2/2     Running            0          70m
capi-kubeadm-control-plane-system   capi-kubeadm-control-plane-controller-manager-7c8ccbb5f-gvmt8   1/2     CrashLoopBackOff   11         70m
capi-system                         capi-controller-manager-7587477b8b-v2hsk                        2/2     Running            0          70m
capi-webhook-system                 capi-controller-manager-847f8c589c-4m826                        2/2     Running            1          70m
capi-webhook-system                 capi-kubeadm-bootstrap-controller-manager-b68c9fb8d-9dp96       2/2     Running            1          70m
capi-webhook-system                 capi-kubeadm-control-plane-controller-manager-b5769bffb-v6hmn   2/2     Running            0          70m
capi-webhook-system                 capm3-controller-manager-76cc77ff64-pjgvd                       2/2     Running            0          70m
capm3-system                        capm3-controller-manager-c5cb7997-24f4q                         2/2     Running            0          70m
cert-manager                        cert-manager-544d659678-45mwv                                   1/1     Running            0          75m
cert-manager                        cert-manager-cainjector-64c9f978d7-hsbx2                        1/1     Running            0          75m
cert-manager                        cert-manager-webhook-5855bb8c8c-sz7s6                           1/1     Running            0          75m
kube-system                         coredns-66bff467f8-l6cw6                                        1/1     Running            0          80m
kube-system                         etcd-minikube                                                   1/1     Running            1          80m
kube-system                         kube-apiserver-minikube                                         1/1     Running            1          80m
kube-system                         kube-controller-manager-minikube                                1/1     Running            1          80m
kube-system                         kube-proxy-57qsx                                                1/1     Running            0          80m
kube-system                         kube-scheduler-minikube                                         1/1     Running            1          80m
kube-system                         storage-provisioner                                             1/1     Running            1          80m
metal3                              metal3-baremetal-operator-65cdd76675-2p7qf                      1/1     Running            0          76m
metal3                              metal3-ironic-769dd8fcdc-xck8g                                  6/6     Running            0          76m
```

**Comment by Xenwar** [comment]
```
ubuntu@upgrade-env-anwar:~$ kubectl describe pods capi-kubeadm-control-plane-controller-manager-7c8ccbb5f-gvmt8 -n capi-kubeadm-control-plane-system
Name:         capi-kubeadm-control-plane-controller-manager-7c8ccbb5f-gvmt8
Namespace:    capi-kubeadm-control-plane-system
Priority:     0
Node:         minikube/192.168.39.180
Containers:
  ......
  manager:
    Container ID:  docker://5ce314ef0c31c55ad1faf66921912c15fc243dd6c9bcc6d0e800fd07baf0b292
    Image:         us.gcr.io/k8s-artifacts-prod/cluster-api/kubeadm-control-plane-controller:v0.3.0
    Image ID:      docker-pullable://us.gcr.io/k8s-artifacts-prod/cluster-api/kubeadm-control-plane-controller@sha256:b33cbe3e58aba8c14932257faeb652e5d9a5e396b34412ff51b3f62db8a95758
    Port:          <none>
    Host Port:     <none>
    Command:
      /manager
    Args:
      --metrics-addr=127.0.0.1:8080
      --enable-leader-election
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Fri, 17 Jul 2020 13:19:07 +0000
      Finished:     Fri, 17 Jul 2020 13:19:27 +0000
    Ready:          False
    Restart Count:  11
    Environment:    <none>
 
Events:
  Type     Reason   Age                   From               Message
  ----     ------   ----                  ----               -------
  Warning  BackOff  109s (x153 over 38m)  kubelet, minikube  Back-off restarting failed container
ubuntu@upgrade-env-anwar:~$ 
```

**Comment by detiber** [comment]
@Xenwar is there a reason you are running v0.3.0 for the KubeadmControlPlane provider? Can you try to replicate with the latest released version (v0.3.7)? There have been a lot of bug fixes since v0.3.0 was released.

**Comment by Xenwar** [comment]
@detiber OoO now.
Will try on Monday

**Comment by Xenwar** [comment]
> @Xenwar is there a reason you are running v0.3.0 for the KubeadmControlPlane provider? Can you try to replicate with the latest released version (v0.3.7)? There have been a lot of bug fixes since v0.3.0 was released.

Sorry, did not pay attention earlier.
The reason I am using v0.3.0 is that we have a use case for upgrading controllers (CAPI and CAPM3) and boot disk image of bare metal servers. So, we needed to fix the version of the controllers at v0.3.0.

**Comment by ncdc** [comment]
This is a duplicate of #2633. If you must perform an upgrade using v0.3.0, you must ensure you have a non-nil kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.

/close
